File: lib/porteiro.py
"""
lib/porteiro.py — a regra de quem pode ir ao ar.

Lógica PURA, sem I/O: recebe o metadata de um dossiê e diz se publicar exige
confirmação extra do operador. Consumida pelo sync_to_cloud (snapshot) e
espelhada em web/lib/porteiro.ts (portal). As duas metades existem de
propósito e as duas são testadas — testar só uma deixa a outra livre pra vazar.

Regra da casa: falhar FECHADO. Na dúvida, bloqueia.
"""
from __future__ import annotations

import unicodedata

# Tags que o Analista usa quando a apuração não fechou. Comparadas JÁ normalizadas.
TAGS_BLOQUEIO = {
    "nao-verificado",
    "apuracao-incompleta",
    "pendente",
    "nao-confirmado",
    "tema-sensivel",
    "rumor",
}


def normaliza_tag(tag: str) -> str:
    """'tema sensível' -> 'tema-sensivel'.

    Minúscula, sem acento, espaço e underscore viram hífen. As tags vêm do
    modelo em texto livre; sem isso a comparação não casa.
    """
    sem_acento = "".join(
        c for c in unicodedata.normalize("NFD", str(tag))
        if unicodedata.category(c) != "Mn"
    )
    return "-".join(sem_acento.lower().split()).replace("_", "-")


def motivo_bloqueio(meta: dict) -> str | None:
    """Por que publicar este dossiê exige confirmação extra — ou None se está limpo.

    A string devolvida vai pra tela do operador, então cita a tag ORIGINAL
    (acentuada), não a normalizada.
    """
    if str((meta or {}).get("confianca", "")).lower() == "baixa":
        return "confiança baixa"
    for tag in (meta or {}).get("tags") or []:
        if normaliza_tag(tag) in TAGS_BLOQUEIO:
            return f"tag de bloqueio: {tag}"
    return None

File: web/lib/porteiro.ts
// Espelho TypeScript de lib/porteiro.py. As duas metades existem de propósito:
// o Python decide o que entra no snapshot, o TS decide o que o portal serve.
// Testar só uma deixa a outra livre pra vazar. Sem imports — lógica pura.
//
// Regra da casa: falhar FECHADO. Na dúvida, não vai ao ar.

export const TAGS_BLOQUEIO = new Set([
  "nao-verificado",
  "apuracao-incompleta",
  "pendente",
  "nao-confirmado",
  "tema-sensivel",
  "rumor",
]);

/** "tema sensível" -> "tema-sensivel" */
export function normalizaTag(tag: string): string {
  return String(tag)
    .normalize("NFD")
    .replace(/\p{Mn}/gu, "")
    .toLowerCase()
    .trim()
    .split(/\s+/)
    .join("-")
    .replace(/_/g, "-");
}

/** Por que publicar exige confirmação extra — ou null se está limpo. */
export function motivoBloqueio(
  meta: { confianca?: string; tags?: string[] } | null | undefined,
): string | null {
  if (!meta) return null;
  if (String(meta.confianca ?? "").toLowerCase() === "baixa") return "confiança baixa";
  for (const tag of meta.tags ?? []) {
    if (TAGS_BLOQUEIO.has(normalizaTag(tag))) return `tag de bloqueio: ${tag}`;
  }
  return null;
}

/** O portal só serve o que o operador promoveu e não arquivou. */
export function podeIrAoAr(d: { status?: string; arquivado?: boolean } | null | undefined): boolean {
  if (!d) return false;
  return d.status === "published" && d.arquivado !== true;
}

/**
 * Devolve AAAA-MM-DD. Aceita ISO e RFC-822 (o WordPress mistura os dois).
 *
 * Espelha lib/dossier_index.normaliza_data. Cortar os 10 primeiros caracteres
 * — o que se fazia antes — transforma "Wed, 22 Apr 2026 11:44:44 +0000" em
 * "Wed, 22 Ap", que ordena ACIMA de qualquer data ISO e sequestra o card de
 * destaque da home.
 */
export function normalizaData(raw: string | null | undefined): string {
  const s = String(raw ?? "").trim();
  if (!s) return "";
  if (/^\d{4}-\d{2}-\d{2}/.test(s)) return s.slice(0, 10);
  const d = new Date(s);
  if (Number.isNaN(d.getTime())) return "";
  return d.toISOString().slice(0, 10);
}
